lib/glossarist/citation.rb

```diff
@@ -1,69 +1,99 @@
-# frozen_string_literal: true
-
-# (c) Copyright 2021 Ribose Inc.
-#
-
 module Glossarist
-  class Citation < Model
+  class Citation < Lutaml::Model::Serializable
     # Unstructured (plain text) reference.
     # @return [String]
-    attr_accessor :text
+    attribute :text, :string
 
     # Source in structured reference.
     # @return [String]
-    attr_accessor :source
+    attribute :source, :string
 
     # Document ID in structured reference.
     # @return [String]
-    attr_accessor :id
+    attribute :id, :string
 
     # Document version in structured reference.
     # @return [String]
-    attr_accessor :version
+    attribute :version, :string
 
     # @return [String]
     # Referred clause of the document.
-    attr_accessor :clause
+    attribute :clause, :string
 
     # Link to document.
     # @return [String]
-    attr_accessor :link
+    attribute :link, :string
 
     # Original ref text before parsing.
     # @return [String]
     # @note This attribute is likely to be removed or reworked in future.
     #   It is arguably not relevant to Glossarist itself.
-    attr_accessor :original
+    attribute :original, :string
 
-    # Whether it is a plain text ref.
-    # @return [Boolean]
-    def plain?
-      (source && id && version).nil?
+    attribute :ref, :string
+
+    yaml do
+      map :id, to: :id, with: { from: :id_from_hash, to: :id_to_hash }
+      map :text, to: :text, with: { from: :text_from_hash, to: :text_to_hash }
+      map :source, to: :source, with: { from: :source_from_hash, to: :source_to_hash }
+      map :version, to: :version, with: { from: :version_from_hash, to: :version_to_hash }
+      map :ref, to: :ref, with: { from: :ref_from_hash, to: :ref_to_hash }
+
+      map :clause, to: :clause
+      map :link, to: :link
+      map :original, to: :original
     end
 
-    # Whether it is a structured ref.
-    # @return [Boolean]
-    def structured?
-      !plain?
+    def ref_from_hash(model, value)
+      model.ref = value
     end
 
-    def to_h
-      {
-        "ref" => ref_to_h,
-        "clause" => clause,
-        "link" => link,
-        "original" => original,
-      }.compact
+    def ref_to_hash(model, doc)
+      doc["ref"] = if model.structured?
+                     ref_hash(model)
+                   else
+                     model.text
+                   end
+    end
```

Comment:
I added a custom method because `ref` can be a string or a hash.

Reply:
Is it possible to dictate only one behavior? Because to support both it’s quite strange.

Reply:
@ronaldtse Sure, Which one do we need to support here?

ref can be like
```yaml
ref: 'ISO 19107:2019'
```

OR

```yaml
ref:
  id: 1234
  source: 'ISO 19107:2019'
  version: 1.3
```

I think we can only support the second format and convert the first one to 

```yaml
ref:
  source: 'ISO 19107:2019'
```